File: common/rpc/CallOutRecord.cpp

#include "CallOutRecord.h"

using namespace zonda::common;

int CallOutRecord::add(const CallOutRecordItem& item)
{
    m_lock.lock();
    //LOG_ERROR("call_id:" << item.rpc_head->call_id);
    m_map[item.rpc_head->call_id] = item;
    m_lock.unlock();
    return 0;
}

int CallOutRecord::remove(int64_t call_id, IMsg* &req)
{
    unordered_map<int64_t, CallOutRecordItem>::iterator it;
    int r = 0;
    m_lock.lock();
    it = m_map.find(call_id);
    if (it != m_map.end())
    {
        //Free the rpc_head, which is the record for thr out req
        req = it->second.req;
        delete it->second.rpc_head;
        m_map.erase(it);
    }
    else
    {
        r = -1;
    }
        
    m_lock.unlock();
    return r;
}

int CallOutRecord::remove(int64_t call_id, bool free_rpc_head)
{
    unordered_map<int64_t, CallOutRecordItem>::iterator it;
    int r = 0;
    m_lock.lock();
    it = m_map.find(call_id);
    if (it != m_map.end())
    {
        //Free the rpc_head, which is the record for thr out req
        if (free_rpc_head) delete it->second.rpc_head;
        m_map.erase(it);
    }
    else
    {
        r = -1;
    }
        
    m_lock.unlock();
    return r;    
}

int CallOutRecord::handle_timeout_items()
{
    unordered_map<int64_t, CallOutRecordItem>::iterator it;
    time_t now = time(NULL);
    m_lock.lock();
    for (it = m_map.begin(); it != m_map.end(); )
    {
        //LOG_ERROR("now:" << now << ", send_time:" << it->second.send_time
        //    << ", timeout:" << it->second.timeout);
        if (now - it->second.send_time >= 30 /*it->second.timeout*/ )
        {
            //We need to make a fake res for the real invoker stage
            //We reuse the rpc_head of the req
            RpcHead* rpc_head = it->second.rpc_head;
            rpc_head->msg_type = 0 - rpc_head->msg_type;
            rpc_head->res_code = RpcResCode::TIMEOUT;
            LOG_INFO("call_id:" << rpc_head->call_id 
                << " times out. biz_handler_id:" 
                << rpc_head->biz_handler_id);
                
            RpcEvent* rpc_event = new RpcEvent(rpc_head);
            rpc_event->set_req(it->second.req);
            m_map.erase(it++);
            
            IEventSink* sink = NULL;
            sink = StageMgr::instance()->get_stage(rpc_head->biz_handler_id)->get_sink();
            
            sink->enqueue(rpc_event);   
            
                      
        }
        else
        {
            ++it;
        }
    }
    m_lock.unlock();
    return 0;
}

size_t CallOutRecord::size()
{
    size_t s = 0;
    m_lock.lock();
    s = m_map.size();
    m_lock.unlock();
    return s;
}

File: common_java/src/main/java/com/alibaba/platform/zonda/common/rpc/CallOutRecord.java
package com.alibaba.platform.zonda.common.rpc;

import java.util.HashMap;
import java.util.Map;
import java.util.Iterator;
import com.alibaba.platform.zonda.logger.lib.AliLoggerFactory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import com.alibaba.platform.zonda.common.seda.*;

class CallOutRecordItem
{
    public RpcHead rpcHead;
    public long sendTime = System.currentTimeMillis()/1000L;
    //The unit is second
    public int timeout;
    public IMsg req = null;
    public IMsg res = null;
    public short resCode = 0;
    public boolean sync = false;
    public boolean timeoutFlag = false;
    public synchronized void wait_res() throws Exception
    {
        if (sync == true && res == null )
        {
            wait(timeout*1000);
            if (res == null)
            {
                //The res has not come back yet, namely timeout
                timeoutFlag = true;
                resCode = (short)RpcHead.RpcResCode.TIMEOUT.ordinal();
            }
        }
    }
    
    public synchronized void notify_res()
    {
        notify();
    }    
}

class CallOutRecord
{
    private Logger logger = AliLoggerFactory.getLogger(Constants.MODULE_NAME, getClass());
    private HashMap<Long, CallOutRecordItem> m_map = new HashMap<Long, CallOutRecordItem>(); 
    public synchronized void add(CallOutRecordItem item)
    {        
        m_map.put(item.rpcHead.callId.get(), item);
    }
    
    public synchronized CallOutRecordItem remove(long callId) throws Exception
    {
        CallOutRecordItem item = m_map.remove(callId);
        if (item == null)
        {
            throw new Exception("Not Found such record");
        }
        
        return item;
    }
    
    public synchronized int handleTimeoutItems()
    {
        Iterator it = m_map.entrySet().iterator();
        Long now = System.currentTimeMillis()/1000L;
        while (it.hasNext()) 
        {
            Map.Entry entry = (Map.Entry) it.next();
            Long callId = (Long)entry.getKey();
            CallOutRecordItem item = (CallOutRecordItem)entry.getValue();
            
            if (now - item.sendTime < 30)
            {
                continue;
            }
            
            
            if (item.sync == true)
            {
                if (item.timeoutFlag == true)
                {
                    //The call thread has returen the timeout
                    //Here only need to remove the item

                }
                else
                {
                    item.resCode = (short)RpcHead.RpcResCode.SEND_FAIL.ordinal();
                    item.notify_res();                
                }
                
                it.remove();
                continue;
            }
            
            
            //For the async call, need to make a fake res for the real invoker stage
            RpcHead rpcHead = item.rpcHead;
            rpcHead.msgType.set(0 - rpcHead.msgType.get());
            rpcHead.resCode.set((short)RpcHead.RpcResCode.TIMEOUT.ordinal());
            logger.info("call_id:" + rpcHead.callId.get()
                + " times out. biz_handler_id:"
                + rpcHead.bizHandlerId.get());

            RpcEvent rpcEvent = new RpcEvent(rpcHead);
            rpcEvent.setReq(item.req);
            it.remove();

            IEventSink sink = null;
            sink = StageMgr.getInstance().getStage(rpcHead.bizHandlerId.get()).getSink();

            sink.enqueue(rpcEvent);            
        } 
        
        return 0;
    }
    
    public synchronized int size()
    {
        return m_map.size();
    }
    

    
    public static void main(String[] args)
    {
        CallOutRecord callOutRecord = new CallOutRecord();
        CallOutRecordItem item1 = new CallOutRecordItem();
        CallOutRecordItem item2 = new CallOutRecordItem();
        
        item1.rpcHead = new RpcHead();
        item1.rpcHead.callId.set(1334039910544001L);
        callOutRecord.add(item1);
        item2.rpcHead = new RpcHead();
        item2.rpcHead.callId.set(1334039910544002L);
        callOutRecord.add(item2);
        
        try
        {
            callOutRecord.remove(1334039910544001L);
            callOutRecord.remove(1334039910544002L);
        }
        catch(Exception e)
        {
            System.out.println(e.getMessage());
            e.printStackTrace();
        }
        
    }   
}